bodo/pandas/_pipeline.cpp

```diff
@@ -2,36 +2,69 @@
 
 #include "physical/result_collector.h"
 
+/*
+ * This has to be a recursive routine.  Each operator in the pipeline could
+ * say that it HAVE_MORE_OUTPUT in which case we need to call it again for the
+ * same input batch (or an empty input batch) once we've call the subsequent
+ * nodes in the pipeline to process the first set of output.  Each operator
+ * could theoretically require multiple (different) iterations in this manner.
+ */
+bool Pipeline::midPipelineExecute(unsigned idx,
+                                  std::shared_ptr<table_info> batch) {
+    // Terminate the recursion when we have processed all the operators
+    // and only have the sink to go which cannot HAVE_MORE_OUTPUT.
+    if (idx >= between_ops.size()) {
+        sink->ConsumeBatch(batch);
+        return false;
+    } else {
+        // Get the current operator.
+        std::shared_ptr<PhysicalSourceSink>& op = between_ops[idx];
+        while (true) {
+            // Process this batch with this operator.
+            std::pair<std::shared_ptr<table_info>, OperatorResult> result =
+                op->ProcessBatch(batch);
+            OperatorResult op_result = result.second;
+
+            // Execute subsequent operators and If any of them said that
+            // no more output is needed or the current operator knows no
+            // more output is needed then return true to terminate the pipeline.
+            if (midPipelineExecute(idx + 1, result.first) ||
+                op_result == OperatorResult::FINISHED) {
```

Comment:
The `op_result == OperatorResult::FINISHED` check has to be first to short circuit? Otherwise early termination doesn't happen.

Reply:
No, I think this order is correct.  You still need to propagate the current batch through the rest of the pipeline to get to the sink.  So, "finished" means that we don't need to generate more output AFTER THIS ONE but the current one needs to go through the pipeline to get to the output.